### ISSUE TITLE
Weights from next sentence prediction

### ISSUE THREAD

**Issue by pierrerappolt** [google-research/bert#370]
Hi, are the final output weights available which were used to predict next sentence (the feed forward layer weights and biases on the CLS token)?

**Comment by pierrerappolt** [comment]
Is it these?
```
cls/seq_relationship/output_bias (DT_FLOAT) [2]
cls/seq_relationship/output_weights (DT_FLOAT) [2,768]
```

**Comment by Alexadar** [comment]
Same question here

**Comment by pierrerappolt** [comment]
@Alexadar I am pretty sure they are saved in `bert_model.ckpt` under the variable names I listed above.

**Comment by hsm207** [comment]
@Alexadar yes, I can confirm those are the correct variable names because of the following snippet from run_pretraining.py:

```
def get_next_sentence_output(bert_config, input_tensor, labels):
  """Get loss and log probs for the next sentence prediction."""

  # Simple binary classification. Note that 0 is "next sentence" and 1 is
  # "random sentence". This weight matrix is not used after pre-training.
  with tf.variable_scope("cls/seq_relationship"):
    output_weights = tf.get_variable(
        "output_weights",
        shape=[2, bert_config.hidden_size],
        initializer=modeling.create_initializer(bert_config.initializer_range))
    output_bias = tf.get_variable(
        "output_bias", shape=[2], initializer=tf.zeros_initializer())

    logits = tf.matmul(input_tensor, output_weights, transpose_b=True)
    logits = tf.nn.bias_add(logits, output_bias)
    log_probs = tf.nn.log_softmax(logits, axis=-1)
    labels = tf.reshape(labels, [-1])
    one_hot_labels = tf.one_hot(labels, depth=2, dtype=tf.float32)
    per_example_loss = -tf.reduce_sum(one_hot_labels * log_probs, axis=-1)
    loss = tf.reduce_mean(per_example_loss)
    return (loss, per_example_loss, log_probs)
```

**Comment by Alexadar** [comment]
Thanks! Any idea how to use vector output from extract_features.py to get next/notNext prediction? I trying to figure out how to use next sentence classifier in pretrained model.

**Comment by Alexadar** [comment]
> @Alexadar Have you considered modifying line 643 in run_classifier.py so that the returned logits are those from get_next_sentence_output()?

Tried for a few weeks. My skills are low. I use pytorch pretrained bert for my home project because i have no idea how to tame tensorflow code :D. If anyone will make eval code for next sentence prediction - will appreciate.

**Comment by Alexadar** [comment]
> @Alexadar Have you considered modifying line 643 in run_classifier.py so that the returned logits are those from get_next_sentence_output()?

Yes, i did it. Each time different results because logits are a product of tensors with initialisers, get_next_sentence_output used for training only and apparently not useful for evaluation. Guys, i wonder how to calculate next sentence weight from pooled layer?

Pytorch implementation of bert model done it in a neat way:

https://github.com/huggingface/pytorch-pretrained-BERT/blob/master/pytorch_pretrained_bert/modeling.py#L572

`class BertOnlyNSPHead(nn.Module):
    def __init__(self, config):
        super(BertOnlyNSPHead, self).__init__()
        self.seq_relationship = nn.Linear(config.hidden_size, 2)

    def forward(self, pooled_output):
        seq_relationship_score = self.seq_relationship(pooled_output)
        return seq_relationship_score`

I wonder, can anyone confirm its correct approach to do it? (nn.Linear solver)

**Comment by Alexadar** [comment]
To test how to deal with next sentence predictions, I created a code to test inference for nextSentence prediction.
The goal is to extract weight between sentences, in a context of sequential relation
Predictions are relatively good, but each time different, i assume because pooled layer was initialised with initialiser, as well as output_weights and output_bias vars

Help me please figure out if this is a correct way to grab weights between sentences, and if not, what to do. Thank you
```

from __future__ import absolute_import, division, print_function, unicode_literals

import os
import sys
import modeling
import optimization
import tensorflow as tf

print(tf.VERSION)

modelPath = "..."

bert_config = modeling.BertConfig.from_json_file(modelPath + "bert_config.json")

# {"label_ids": 2,"input_ids": [101, 1573, 117, 1115, 1116, 1725, 27023, 1116, 1132, 9245, 1116, 102, 140, 8741, 1162, 1108, 1276, 1120, 1139, 6023, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "input_mask": [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "segment_ids": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}

input_ids = tf.convert_to_tensor([[101, 1573, 117, 1115, 1116, 1725, 27023, 1116, 1132, 9245, 1116, 102, 140, 8741, 1162, 1108, 1276, 1120, 1139, 6023, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=tf.int32)
input_mask = tf.convert_to_tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=tf.int32)
segment_ids = tf.convert_to_tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=tf.int32)

model = modeling.BertModel(
    config=bert_config,
    is_training=False,
    input_ids=input_ids,
    input_mask=input_mask,
    token_type_ids=segment_ids,
    use_one_hot_embeddings=False)

tvars = tf.trainable_variables()

(assignment_map, initialized_variable_names
 ) = modeling.get_assignment_map_from_checkpoint(tvars, modelPath + "bert_model.ckpt")
tf.train.init_from_checkpoint(modelPath + "bert_model.ckpt", assignment_map)

def get_next_sentence_output(bert_config, input_tensor, labels):
    """Get loss and log probs for the next sentence prediction."""

    # Simple binary classification. Note that 0 is "next sentence" and 1 is
    # "random sentence". This weight matrix is not used after pre-training.
    with tf.variable_scope("cls/seq_relationship"):
        output_weights = tf.get_variable(
            "output_weights",
            shape=[2, bert_config.hidden_size],
            initializer=tf.truncated_normal_initializer(stddev=0.02))
        output_bias = tf.get_variable(
            "output_bias", shape=[2], initializer=tf.zeros_initializer())

        logits = tf.matmul(input_tensor, output_weights, transpose_b=True)
        logits = tf.nn.bias_add(logits, output_bias)
        log_probs = tf.nn.log_softmax(logits, axis=-1)
        probs = tf.nn.softmax(logits, axis=-1)
        labels = tf.reshape(labels, [-1])
        one_hot_labels = tf.one_hot(labels, depth=2, dtype=tf.float32)
        per_example_loss = -tf.reduce_sum(one_hot_labels * log_probs, axis=-1)
        loss = tf.reduce_mean(per_example_loss)
        return (loss, per_example_loss, log_probs, logits, probs)


next_sentence_labels =  tf.convert_to_tensor([[0]], tf.int64)

pooled = model.get_pooled_output()
softmax = tf.nn.softmax(pooled)

(next_sentence_loss, next_sentence_example_loss,
 next_sentence_log_probs, logits, probs) = get_next_sentence_output(
    bert_config, pooled,next_sentence_labels)

with tf.compat.v1.Session() as sess:
    init = tf.global_variables_initializer()
    sess.run(init)
    print(next_sentence_log_probs.eval())
    print(probs.eval())
```

**Comment by lixiangtnt** [comment]
@Alexadar
I think you can modify run_classifier.py via the reference of def get_next_sentence_output(bert_config, input_tensor, labels).

Simply adding "with tf.variable_scope("cls/seq_relationship"): "  one top of the 
    output_weights = tf.get_variable(
            "output_weights",
           ....
        output_bias = tf.get_variable(
            "output_bias", shape=[2], initializer=tf.zeros_initializer())
Replacing the "input_tensor"  in  " logits = tf.matmul(input_tensor, output_weights, transpose_b=True)" into "output_layer"
